File: inventario/src/CX/SesionUsuario.java
package CX;

public class SesionUsuario {
    private static SesionUsuario instancia;

    private Long usuarioId;
    private String nombreUsuario;
    private String rol;
    private String permisos;

    private SesionUsuario() {}

    public static SesionUsuario getInstancia() {
        if (instancia == null) {
            instancia = new SesionUsuario();
        }
        return instancia;
    }

    public void iniciarSesion(long id, String nombre, String rol, String permisos) {
        this.usuarioId = id;
        this.nombreUsuario = nombre;
        this.rol = rol;
        this.permisos = permisos;
    }

    public void cerrarSesion() {
        this.usuarioId = null;
        this.nombreUsuario = null;
        this.rol = null;
        this.permisos = null;
    }

    public Long getUsuarioId() {
        return usuarioId;
    }

    public String getNombreUsuario() {
        return nombreUsuario;
    }

    public String getRol() {
        return rol;
    }

    public String getPermisos() {
        return permisos;
    }

    public void setPermisos(String permisos) {
        this.permisos = permisos;
    }

    private String normalizar(String s) {
        if (s == null) return "";
        return s.toUpperCase()
                .replace("Á", "A").replace("É", "E").replace("Í", "I").replace("Ó", "O").replace("Ú", "U")
                .replaceAll("[^A-Z0-9:,]", ""); // Eliminar caracteres extraños o acentos no mapeados
    }

    public boolean tienePermiso(String modulo) {
        String p = normalizar(permisos);
        String m = normalizar(modulo);
        
        int idx = p.indexOf(m + ":");
        if (idx != -1) {
            int endIdx = p.indexOf(",", idx);
            String sub = (endIdx == -1) ? p.substring(idx) : p.substring(idx, endIdx);
            String actionsPart = sub.substring(sub.indexOf(":") + 1);
            return !actionsPart.isEmpty(); // Si tiene cualquier letra (C, E, D, X), tiene permiso de entrar
        }
        
        if ("Admin".equalsIgnoreCase(rol) || "Administrador".equalsIgnoreCase(rol)) return true;
        return p.contains(m);
    }

    public boolean tienePermiso(String modulo, String accion) {
        if (permisos == null && !("Admin".equalsIgnoreCase(rol) || "Administrador".equalsIgnoreCase(rol))) return false;
        
        String p = normalizar(permisos);
        String m = normalizar(modulo);
        String a = "";
        switch (accion.toUpperCase()) {
            case "VER": return tienePermiso(modulo); // Caso especial: Ver depende de tener cualquier permiso
            case "CREAR": a = "C"; break;
            case "EDITAR": a = "E"; break;
            case "ELIMINAR": a = "D"; break;
            case "EXPORTAR": a = "X"; break;
            default: a = accion.toUpperCase().substring(0, 1);
        }

        // Prioridad 1: Si existe una definición específica para el módulo (ej: PRODUCTOS:VCE)
        int idx = p.indexOf(m + ":");
        if (idx != -1) {
            int endIdx = p.indexOf(",", idx);
            String sub = (endIdx == -1) ? p.substring(idx) : p.substring(idx, endIdx);
            String actionsPart = sub.substring(sub.indexOf(":") + 1);
            return actionsPart.contains(a);
        }
        
        // Prioridad 2: Si es Administrador y no hay restricción específica, tiene permiso
        if ("Admin".equalsIgnoreCase(rol) || "Administrador".equalsIgnoreCase(rol)) return true;
        
        // Prioridad 3: Formato antiguo o permiso general por nombre
        return p.contains(m);
    }
}

File: inventario_python/core/sesion_usuario.py
import re


class SesionUsuario:
    _instancia = None

    def __new__(cls):
        if cls._instancia is None:
            cls._instancia = super().__new__(cls)
            cls._instancia.usuario_id = None
            cls._instancia.nombre_usuario = None
            cls._instancia.rol = None
            cls._instancia.permisos = None
        return cls._instancia

    def iniciar_sesion(self, id, nombre, rol, permisos):
        self.usuario_id = id
        self.nombre_usuario = nombre
        self.rol = rol
        self.permisos = permisos

    def cerrar_sesion(self):
        self.usuario_id = None
        self.nombre_usuario = None
        self.rol = None
        self.permisos = None

    def _normalizar(self, s):
        if not s:
            return ""
        s = s.upper()
        replacements = {"Á": "A", "É": "E", "Í": "I", "Ó": "O", "Ú": "U"}
        for k, v in replacements.items():
            s = s.replace(k, v)
        s = re.sub(r"[^A-Z0-9:,]", "", s)
        return s

    def tiene_permiso(self, modulo):
        p = self._normalizar(self.permisos)
        m = self._normalizar(modulo)
        idx = p.find(m + ":")
        if idx != -1:
            end = p.find(",", idx)
            sub = p[idx:] if end == -1 else p[idx:end]
            actions = sub[sub.index(":") + 1:]
            return bool(actions)
        if self.rol and self.rol.lower() in ("admin", "administrador"):
            return True
        return m in p

    def tiene_permiso_accion(self, modulo, accion):
        if self.permisos is None and not (self.rol and self.rol.lower() in ("admin", "administrador")):
            return False
        p = self._normalizar(self.permisos)
        m = self._normalizar(modulo)
        a_map = {"VER": None, "CREAR": "C", "EDITAR": "E", "ELIMINAR": "D", "EXPORTAR": "X"}
        a = a_map.get(accion.upper(), accion.upper()[:1])
        if a is None:
            return self.tiene_permiso(modulo)
        idx = p.find(m + ":")
        if idx != -1:
            end = p.find(",", idx)
            sub = p[idx:] if end == -1 else p[idx:end]
            actions = sub[sub.index(":") + 1:]
            return a in actions
        if self.rol and self.rol.lower() in ("admin", "administrador"):
            return True
        return m in p
